File: src/leetCodeProblems/BinaryTree/BinaryTreeVerticalOrderTraversal314.java
package leetCodeProblems.BinaryTree;

/**
 * LeetCode - https://leetcode.com/problems/binary-tree-vertical-order-traversal/
 * 
 */
import java.util.*;

public class BinaryTreeVerticalOrderTraversal314 {

	int min = 0;
    int max = 0;
    
    public List<List<Integer>> result = new ArrayList<List<Integer>>();
    
    private void findMinMax(TreeNode node, int currentVerticalOrder) {
        
        if (node == null) {
            return;
        }
        
        min = Math.min(min, currentVerticalOrder);
        max = Math.max(max, currentVerticalOrder);
        
        findMinMax(node.left, currentVerticalOrder-1);
        findMinMax(node.right, currentVerticalOrder+1);
    }
    
    private void bfsTraversal(TreeNode root) {
    	Queue<TreeVNode> queue = new LinkedList<TreeVNode>();
        
        TreeVNode nodeV = new TreeVNode(root, 0);
        queue.add(nodeV);
        
        TreeVNode currentVNode;
        TreeNode currentNode; 
        int currentNodeVOrder;
        
        TreeVNode currentLeftVNode;
        TreeVNode currentRightVNode;
        
        while (!queue.isEmpty()) {
            
            currentVNode = queue.remove();
            currentNode = currentVNode.node;
            currentNodeVOrder = currentVNode.vOrder;
            
            if ( currentNode != null) {
                
                int currentNodeVOrderArrayListIndex = currentNodeVOrder-min;
                result.get(currentNodeVOrderArrayListIndex).add(currentNode.val);
                
                if (currentNode.left != null) {
                	currentLeftVNode = new TreeVNode(currentNode.left, currentNodeVOrder-1);
                    queue.add(currentLeftVNode);
                }
                
                if (currentNode.right != null) {
                	currentRightVNode = new TreeVNode(currentNode.right, currentNodeVOrder+1);
                    queue.add(currentRightVNode);
                }
            }
        }
        
        return;
    }
    
    // Driver method
    public List<List<Integer>> verticalOrder(TreeNode root) {
        
        if (root == null) {
            return new ArrayList<>();    
        }
        
        findMinMax(root, 0);
        
        int numberOfVTraversalRows = max - min;
        
        for(int i = 0; i <= numberOfVTraversalRows; i++) {
            result.add(new ArrayList<Integer>());
        }
        
        bfsTraversal(root);
        
        return result;
        
    }
    
    static class TreeNode {
    	int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    
    static class TreeVNode {
        TreeNode node;
        int vOrder;
        
        TreeVNode(TreeNode node, int vOrder) {
            this.node = node;
            this.vOrder = vOrder;
        }
    }
    
    public static void main(String[] args) {
        //System.out.println("Hello World!");
        
        /*TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.left.left = new TreeNode(4);
        root.left.right = new TreeNode(5);
        root.right.left = new TreeNode(6);
        root.right.right = new TreeNode(7);*/
        
        TreeNode root = new TreeNode(3);
        root.left = new TreeNode(9);
        root.right = new TreeNode(8);
        root.left.left = new TreeNode(4);
        root.left.right = new TreeNode(0);
        root.right.left = new TreeNode(1);
        root.right.right = new TreeNode(7);
        
        root.left.right.left = new TreeNode(5);
        root.left.right.right = new TreeNode(2);
        
        
        BinaryTreeVerticalOrderTraversal314 obj = new BinaryTreeVerticalOrderTraversal314();
        
        List<List<Integer>> result = obj.verticalOrder(root);
        
        for(int i=0; i < result.size(); i++) {
            System.out.println(result.get(i));
        }
        
    }
}

File: src/leetCodeProblems/BinaryTree/BinaryTreeVerticalOrderTraversal314.go
package main

/*
- LeetCode - https://leetcode.com/problems/binary-tree-vertical-order-traversal/
*/
import (
	"log"
	"sort"
)

type TreeVNode struct {
	node   *TreeNode
	vOrder int
}

type tQueue []TreeVNode

func (q *tQueue) pop() TreeVNode {
	out := (*q)[0]
	*q = (*q)[1:]
	return out
}

func (q *tQueue) push(t TreeVNode) {
	*q = append(*q, t)
}

func (q *tQueue) isEmpty() bool {
	return len(*q) == 0
}

func findMinMaxVOrder(node *TreeNode, current int, min *int, max *int) {

	if node == nil {
		return
	}

	leftVOrder := current
	rightVOrder := current

	if node.Left != nil {
		leftVOrder--
	}

	if node.Right != nil {
		rightVOrder++
	}

	if *min > leftVOrder {
		*min = leftVOrder
	}

	if *max < rightVOrder {
		*max = rightVOrder
	}

	findMinMaxVOrder(node.Left, leftVOrder, min, max)
	findMinMaxVOrder(node.Right, rightVOrder, min, max)
}

func verticalOrder(root *TreeNode) [][]int {
	min, max := 0, 0

	findMinMaxVOrder(root, 0, &min, &max)

	total := max - min + 1

	log.Println(min, max, total)

	queue := new(tQueue)
	queue.push(TreeVNode{root, 0})

	out := make([][]int, total)

	for i := range out {
		out[i] = []int{}
	}

	for !queue.isEmpty() {
		cTreeVNode := queue.pop()
		outIndex := cTreeVNode.vOrder - min
		log.Println(outIndex)
		out[outIndex] = append(out[outIndex], cTreeVNode.node.Val)

		if cTreeVNode.node.Left != nil {
			queue.push(TreeVNode{cTreeVNode.node.Left, cTreeVNode.vOrder - 1})
		}

		if cTreeVNode.node.Right != nil {
			queue.push(TreeVNode{cTreeVNode.node.Right, cTreeVNode.vOrder + 1})
		}
	}

	for i := range out {
		sort.Ints(out[i])
	}
	return out

}

func main() {
	// root := new(TreeNode) //Returns pointer to TreeNode object
	// root.Val = 3
	// root.Left = new(TreeNode)
	// root.Left.Val = 9
	// root.Right = new(TreeNode)
	// root.Right.Val = 20
	// root.Right.Left = new(TreeNode)
	// root.Right.Left.Val = 15
	// root.Right.Right = new(TreeNode)
	// root.Right.Right.Val = 7

	// root := new(TreeNode) //Returns pointer to TreeNode object
	// root.Val = 1
	// root.Left = new(TreeNode)
	// root.Left.Val = 2
	// root.Left.Left = new(TreeNode)
	// root.Left.Left.Val = 4
	// root.Left.Right = new(TreeNode)
	// root.Left.Right.Val = 5

	// root.Right = new(TreeNode)
	// root.Right.Val = 3
	// root.Right.Left = new(TreeNode)
	// root.Right.Left.Val = 6
	// root.Right.Right = new(TreeNode)
	// root.Right.Right.Val = 7

	root := new(TreeNode) //Returns pointer to TreeNode object
	root.Val = 0
	root.Left = new(TreeNode)
	root.Left.Val = 1
	root.Left.Right = new(TreeNode)
	root.Left.Right.Val = 2
	root.Left.Right.Left = new(TreeNode)
	root.Left.Right.Left.Val = 6
	root.Left.Right.Right = new(TreeNode)
	root.Left.Right.Right.Val = 3
	root.Left.Right.Right.Right = new(TreeNode)
	root.Left.Right.Right.Right.Val = 4
	root.Left.Right.Right.Right.Right = new(TreeNode)
	root.Left.Right.Right.Right.Right.Val = 5

	log.Println(verticalTraversal(root))
}
